### PR TITLE
WIP: Improve DiagnosticsEngine support in Workspace

### PR DESCRIPTION
**Pull request by hartbit**

### CODE REVIEW

File: Sources/Workspace/Workspace.swift [hunk 1]
```diff
@@ -290,9 +287,18 @@ public class Workspace {
...
+            try fileSystem.createDirectory(repositoryManager.path, recursive: true)
+        } catch {
+            engine.emit(error)
+        }
+        
+        do {
+            try fileSystem.createDirectory(checkoutsPath, recursive: true)
+        } catch {
+            engine.emit(error)
+        }
```
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: I don't think we need separate do catch here, a failure here would mean this method needs to be retried anyway.

File: Sources/Workspace/Workspace.swift [hunk 2]
```diff
@@ -302,27 +308,29 @@ public class Workspace {
             repositoryManager.path,
             checkoutsPath,
             try managedDependencies.load().statePath
-        ].map { path in
+        ].map({ path in
             // Assert that these are present inside data directory.
             assert(path.parentDirectory == dataPath)
             return path.basename
-        })
+        }))
```
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: nit: unrelated changes should usually be a separate PR/commit

File: Sources/Workspace/Workspace.swift [hunk 3]
```diff
@@ -302,27 +308,29 @@ public class Workspace {
...
         for name in try fileSystem.getDirectoryContents(dataPath) {
             guard !protectedAssets.contains(name) else { continue }
             fileSystem.removeFileTree(dataPath.appending(RelativePath(name)))
         }
     }
 
     /// Resets the entire workspace by removing the data directory.
-    public func reset() throws {
+    public func reset(engine: DiagnosticsEngine) {
         try managedDependencies.load().reset()
```
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: There is an overload to do this via engine.

File: Sources/Workspace/Workspace.swift [hunk 4]
```diff
@@ -340,13 +349,15 @@ public class Workspace {
     public func edit(
         dependency: ManagedDependency,
         packageName: String,
+        engine: DiagnosticsEngine,
```
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: lets use `diagnostics` instead of engine as it isn't very good name. We need to fix this at other places too (but thats a separate thing)

File: Sources/Workspace/Workspace.swift [hunk 5]
```diff
@@ -366,15 +377,19 @@ public class Workspace {
...
+                    path: destination,
+                    destinationPackage: manifest.name,
+                    expectedPackage: packageName))
+                return
             }
+            
             // Emit warnings for branch and revision, if they're present.
             if let checkoutBranch = checkoutBranch {
-                delegate.warning(message: "not checking out branch '\(checkoutBranch)' for dependency '\(packageName)'")
+                engine.emit(warning: "not checking out branch '\(checkoutBranch)' for dependency '\(packageName)'")
```
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: What is this overload? We should probably have something like `emit.(WorkspaceWarning("..."))` here

File: Sources/Workspace/Workspace.swift (hunk 4 above)
Comment (hartbit) [inline Sources/Workspace/Workspace.swift]: Do you me to do that in this PR? Seems appropriate.

File: Sources/Workspace/Workspace.swift (hunk 3 above)
Comment (hartbit) [inline Sources/Workspace/Workspace.swift]: Where is that overload?

File: Sources/Workspace/Workspace.swift (hunk 5 above)
Comment (hartbit) [inline Sources/Workspace/Workspace.swift]: This overload does not exist yet. But I was planning on something like: `emit.(WorkspaceWarning.unusedCheckoutBranchArgument)` and define the text in the associated diagnostic ID.

File: Sources/Workspace/Workspace.swift (hunk 3 above)
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: https://github.com/apple/swift-package-manager/blob/master/Sources/Workspace/Workspace.swift#L1226

File: Sources/Workspace/Workspace.swift (hunk 4 above)
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: Yeah I mean use "diagnostics" instead of "engine" for the new additions

File: Sources/Workspace/Workspace.swift (hunk 5 above)
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: That makes sense
